### PR TITLE
ipq40xx: add support for EnGenius EMR5000

### PR DESCRIPTION
**Pull request by Yen-Ting-Shen**
SOC:     IPQ4019 / QCA Dakota
        CPU:     Quad-Core ARMv7 Processor rev 5 (v7l) Cortex-A7
        DRAM:    256 MiB
        FLASH:   NAND 128 MiB
        ETH:     Qualcomm Atheros QCA8072 (2 ports)
        WLAN1:   Qualcomm Atheros QCA4019 2.4GHz 802.11bgn 2:2x2
        WLAN2:   Qualcomm Atheros QCA4019 5GHz(channel 100-165) 802.11a/n/ac 2:2x2
        WLAN3:   Qualcomm Atheros QCA9888 5GHz(channel 36-64) 802.11a/n/ac 2:2x2
        INPUT:   RESET Button
        LEDS:    White, Blue, Red, Orange
        USB:     1x 3.0

        Flash instruction:
        Upload openwrt-ipq40xx-generic-engenius_emr5000-squashfs-nand-factory.bin
        via the firmwary upgrade function of the stock firmware WebUI.

Signed-off-by: Yen-Ting-Shen <frank.shen@senao.com>

### CODE REVIEW

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+	};
+
+	keys {
+			compatible = "gpio-keys";
+
+			reset {
+				label = "reset";
+				gpios = <&tlmm 18 GPIO_ACTIVE_LOW>;
+				linux,code = <KEY_RESTART>;
+			};
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts]: Indent is wrong here.

**Comment by adrianschmutzler** [comment]
Consider adding label_mac as discussed in your other PR ...

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts (hunk 1 above)
Comment (Yen-Ting-Shen) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts]: Fixed.

**Comment by Yen-Ting-Shen** [comment]
> Consider adding label_mac as discussed in your other PR ...

Added.

**Comment by Yen-Ting-Shen** [comment]
removed indents in the commit message.
rebased the PR.

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+	status = "okay";
+};
+
+&wifi0 {
+	status = "ok";
+	qcom,ath10k-calibration-variant = "EnGenius-EMR5000";
+};
+
+&wifi1 {
+	status = "ok";
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts:303]: Please change all cases of "ok" to "okay".

File: target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh [hunk 3]
```diff
@@ -9,6 +9,10 @@ preinit_set_mac_address() {
 		ip link set dev eth0 address $(macaddr_add "$base_mac" +1)
 		ip link set dev eth1 address $(macaddr_add "$base_mac" +3)
 		;;
+	engenius,emr5000)
+		base_mac=$(cat /sys/class/net/eth0/address)
+		ip link set dev eth1 address $(macaddr_add "$base_mac" +3)
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh:14]: Why is eth0 getting a valid address automagically and eth1 isn't?

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+		led-boot = &power;
+		led-failsafe = &power;
+		led-running = &power;
+		led-upgrade = &power;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		power: white {
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts:102]: `led_power: power {`

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+			label = "emr5000:white:power";
+			gpios = <&tlmm 46 GPIO_ACTIVE_HIGH>;
+		};
+
+		blue {
+			label = "emr5000:blue";
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts:108]: So, what are those leds for? You should use a three-membered pattern like for the white led ...

File: target/linux/ipq40xx/image/Makefile [hunk 6]
```diff
@@ -193,6 +193,22 @@ define Device/engenius_eap1300
 endef
 TARGET_DEVICES += engenius_eap1300
 
+define Device/engenius_emr5000
+	$(call Device/FitImage)
+	$(call Device/UbiFit)
+	DEVICE_VENDOR := EnGenius
+	DEVICE_MODEL := EMR5000
+	BOARD_NAME := emr5000
+	DEVICE_DTS := qcom-ipq4019-emr5000
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/image/Makefile:202]: please use
SOC := qcom-ipq4019
instead of DEVICE_DTS.

File: target/linux/ipq40xx/image/Makefile [hunk 7]
```diff
@@ -193,6 +193,22 @@ define Device/engenius_eap1300
...
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	DEVICE_PACKAGES := ath10k-firmware-qca9888 uboot-envtools
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/image/Makefile:205]: please use the matching ath10k-...-ct package instead.

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 8]
```diff
@@ -142,6 +150,10 @@ case "$FIRMWARE" in
 	unielec,u4019-32m)
 		caldata_extract "0:ART" 0x5000 0x2f20
 		;;
+	engenius,emr5000)
+		ath10kcal_extract "0:ART" 0x5000 0x2f20
+		ath10kcal_patch_mac_crc $(macaddr_add "$(cat /sys/class/net/eth0/address)" 1)
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:155]: Have you ever tested what happens if you drop the ath10kcal_patch_mac_crc lines in this file? Will the correct address be assigned anyway (i.e., is the correct address stored in the calibration data?).

**Comment by adrianschmutzler** [comment]
Please update for kernel 5.4. Note that I will add a patch to master soon that will consolidate the DTS files in the target, so maybe wait for that one.

**Pull request closed without merging**